File: ProductExtractor/browser.py
from typing import TYPE_CHECKING


if TYPE_CHECKING:
    from instructions import BrowserInstruction


from shared_types import WebDriver


class Browser:
    def __init__(self, driver: WebDriver):
        self.driver = driver

    def execute[T](self, instruction: "BrowserInstruction[T]") -> T:
        return instruction.accept(self)

    def quit(self):
        self.driver.quit()

File: offer-extractor/browser.ts
import type { WebDriver } from "selenium-webdriver"
import type { Instruction } from "./instructions.ts";

export class Browser {
    driver: WebDriver;

    constructor(driver: WebDriver) {
        this.driver = driver;
    }

    async execute<T>(instruction: Instruction<T>): Promise<T> {
        return await instruction.accept(this);
    }

    async quit() {
        await this.driver.quit()
    }
}
